### PR TITLE
Refactor CUDA implementation of unique and unique_dim to improve performance of unique_dim

### LINKED ISSUE #18405: _unique_dim needs refactor
I am profiling the training of my model which uses 
`tensor.unique(dim=0, sorted=True, return_inverse=True, return_counts=True)` for a tensor of size ~`(15388, 2)`, with a manually merge of https://github.com/pytorch/pytorch/pull/18391. There is only one call of `unique` for each forward pass on my model. Timings for unique_dim is really bad:

```
--------------------------------------------------------------------------------
  autograd profiler output (CUDA mode)
--------------------------------------------------------------------------------
        top 15 events sorted by cpu_time_total

        Because the autograd profiler uses the CUDA event API,
        the CUDA time column reports approximately max(cuda_time, cpu_time).
        Please ignore this output if your code does not use CUDA.

----------------  ---------------  ---------------  ---------------  ---------------  ---------------
Name                     CPU time        CUDA time            Calls        CPU total       CUDA total
----------------  ---------------  ---------------  ---------------  ---------------  ---------------
_unique_dim         1127002.344us    1126993.000us                1    1127002.344us    1126993.000us
_unique_dim         1125635.745us    1125626.000us                1    1125635.745us    1125626.000us
_unique_dim         1120160.511us    1120151.000us                1    1120160.511us    1120151.000us
_unique_dim         1116257.000us    1116249.000us                1    1116257.000us    1116249.000us
_unique_dim         1099370.667us    1099362.000us                1    1099370.667us    1099362.000us
_unique_dim         1085741.313us    1085733.875us                1    1085741.313us    1085733.875us
_unique_dim         1083343.180us    1083333.750us                1    1083343.180us    1083333.750us
_unique_dim         1083208.457us    1083199.000us                1    1083208.457us    1083199.000us
_unique_dim         1082070.081us    1082061.500us                1    1082070.081us    1082061.500us
_unique_dim         1061865.168us    1061857.500us                1    1061865.168us    1061857.500us
_unique_dim          606158.761us     606154.000us                1     606158.761us     606154.000us
index_select           3095.179us       3106.500us                1       3095.179us       3106.500us
index                  2425.817us       2427.875us                1       2425.817us       2427.875us
sum                    1642.662us       1644.250us                1       1642.662us       1644.250us
EluBackward            1629.843us       1630.000us                1       1629.843us       1630.000us
```

_unique and _unique_dim actually need some refactor, https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/Unique.cu
Currently these two operators are implemented seperately, but the logic of these operators should be more or less the same, so they should be combined into a single function such that improves in `_unqiue` (https://github.com/pytorch/pytorch/pull/16145) could also benefit `_unique_dim`.

See also: https://github.com/pytorch/pytorch/issues/15804

@ptrblck @ngimel If you don't mind, I will work on this. But I'm planning to get https://github.com/pytorch/pytorch/pull/18391 merged as is without any refactor and base the refactor on that issue.

Also: cc @ezyang

### PR DESCRIPTION
**Pull request by zasdfgbnm**
Fixes: https://github.com/pytorch/pytorch/issues/18405

This is based on https://github.com/pytorch/pytorch/pull/18391, I suggest first wait until #18391 get merged and then I will rebase.

cc: @ngimel 

# Benchmark

Benchmark is on a tensor of shape `torch.Size([15320, 2])`

## Before:

```python
%timeit a.unique(dim=0, sorted=True, return_inverse=False, return_counts=False)
%timeit a.unique(dim=0, sorted=True, return_inverse=True, return_counts=False)
%timeit a.unique(dim=0, sorted=True, return_inverse=False, return_counts=True)
%timeit a.unique(dim=0, sorted=True, return_inverse=True, return_counts=True)
```
```
200 µs ± 3.14 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
1.04 s ± 21.4 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
1.03 s ± 15 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
1.07 s ± 27.8 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
```python
%timeit a.unique(sorted=True, return_inverse=False, return_counts=False)
%timeit a.unique(sorted=True, return_inverse=True, return_counts=False)
%timeit a.unique(sorted=True, return_inverse=False, return_counts=True)
%timeit a.unique(sorted=True, return_inverse=True, return_counts=True)
```
```
227 µs ± 2.38 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
291 µs ± 2.34 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
249 µs ± 1.56 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
316 µs ± 2.64 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

## After

```python
%timeit a.unique(dim=0, sorted=True, return_inverse=False, return_counts=False)
%timeit a.unique(dim=0, sorted=True, return_inverse=True, return_counts=False)
%timeit a.unique(dim=0, sorted=True, return_inverse=False, return_counts=True)
%timeit a.unique(dim=0, sorted=True, return_inverse=True, return_counts=True)
```
```
187 µs ± 4.67 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
248 µs ± 3.9 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
212 µs ± 906 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
274 µs ± 1.8 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```
```python
%timeit a.unique(sorted=True, return_inverse=False, return_counts=False)
%timeit a.unique(sorted=True, return_inverse=True, return_counts=False)
%timeit a.unique(sorted=True, return_inverse=False, return_counts=True)
%timeit a.unique(sorted=True, return_inverse=True, return_counts=True)
```
```
226 µs ± 763 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
294 µs ± 1.03 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
253 µs ± 3.5 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
323 µs ± 1.52 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

### CODE REVIEW

**Comment by zasdfgbnm** [comment]
OK... This seems to fail on cuda 8...

**Comment by zasdfgbnm** [comment]
@ngimel This should be mostly ready for review now. The only issue is it fails on cuda8, I need to look deep into it.

File: aten/src/ATen/native/cuda/Unique.cu [hunk 1]
```diff
@@ -15,86 +16,103 @@ namespace native{
...
 
-    // sort indices using data
-    thrust::sort(policy, indices_ptr, indices_ptr + indices.numel(),
+    Tensor indices = at::arange(0, num_inp, self.options().dtype(kLong));
+    int64_t *indices_data = indices.data<int64_t>();
+
+    Tensor inverse_indices, counts;
+    int64_t num_out;
+    std::tie(inverse_indices, counts, num_out) = compute_unique(
+      indices_data, num_inp, return_inverse, return_counts, self.options().dtype(kLong),
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/Unique.cu]: compute_unique first argument type is `scalar_t *`, and you are sending `indices_data` which is `int64_t *`, does not look right.

**Review by ngimel — COMMENTED** [review]
The algorithm looks good, except for my type comment, I hope you can figure out cuda 8 stuff (I see you are experimenting with `less` lambda now). Generally, cuda 8 lambda support is spotty, so you might need to define a functor instead of using lambda

**Review by VitalyFedyunin — CHANGES_REQUESTED** [review]
Overall looks good. Please fix CUDA 8.

**Comment by apaszke** [comment]
If it's a CUDA change, can you please add `torch.cuda.synchronize()` at the end of all `%timeit` lines? Otherwise it's not very meaningful.

File: aten/src/ATen/native/cuda/Unique.cu [hunk 2]
```diff
@@ -15,86 +16,103 @@ namespace native{
...
-    auto policy = thrust::cuda::par(allocator).on(stream);
-
-    const Tensor& input = self.contiguous();
-    int64_t num_inp = input.numel();
-    const scalar_t* input_data = input.data<scalar_t>();
-
-    //sort
-    Tensor output = input.clone();
-    output = output.view(-1);
+    Tensor output = self.contiguous().reshape(-1).clone();
```
Comment (fmassa) [inline aten/src/ATen/native/cuda/Unique.cu]: nit: there might be unnecessary copies here.
I believe you can get the same behavior with
```python
self.clone().reshape(-1)
```

File: aten/src/ATen/native/cuda/Unique.cu (hunk 1 above)
Comment (zasdfgbnm) [reply aten/src/ATen/native/cuda/Unique.cu]: @ngimel This is expected. The tensor passed in is the index tensor, and we override the compare operator of it to compare the actual underlying data, and finally use an `index_select` to get the result.

File: aten/src/ATen/native/cuda/Unique.cu (hunk 2 above)
Comment (zasdfgbnm) [reply aten/src/ATen/native/cuda/Unique.cu]: fixed

File: aten/src/ATen/native/cuda/Unique.cu (hunk 1 above)
Comment (ngimel) [reply aten/src/ATen/native/cuda/Unique.cu]: So `scalar_t` is resolved to `int64_t`. Confusing, but ok. However, I don't think this line https://github.com/pytorch/pytorch/pull/18459/files#diff-0ac6387c4c1640433eaf96ce504cb53bR50 works as expected if `data` is in fact indices - the comparison operator has to be overriden here too.

Comment (zasdfgbnm) [reply aten/src/ATen/native/cuda/Unique.cu]: @ngimel I think you are right. There is another problem that the test case is not covering cuda, so we didn't see any error on it. I have fixed the test. But not the implementation itself yet.

Comment (zasdfgbnm) [reply aten/src/ATen/native/cuda/Unique.cu]: @ngimel implementation also fixed now

**Comment by zasdfgbnm** [comment]
close this in favor of https://github.com/pytorch/pytorch/pull/18648
